.openshift-ci/drivers/scripts/kernel-splitter.py

```diff
@@ -0,0 +1,179 @@
+    def shards(self):
+        return self._shards
+
+    @shards.setter
+    def shards(self, s):
+        self._shards = s if s > 0 else 1
+
+    def _dump_shard(self, shard, tasks):
+        output_file = os.path.join(self.output_dir, 'shards', str(shard))
+        with open(output_file, 'w') as f:
+            f.writelines(tasks)
+
+    def _dump_all(self):
+        raw_tasks = [
+            f'{kernel} {module} {driver_type}\n'
+            for kernel, v in self.tasks.items()
+            for module, driver_type in v
+        ]
+
+        with open(os.path.join(self.output_dir, 'all'), 'w') as f:
+            f.writelines(raw_tasks)
+
+    def dump(self):
+        # Create the directory for storing tasks and fill the 'all' tasks file
+        os.makedirs(os.path.join(self.output_dir, 'shards'), exist_ok=True)
+
+        self._dump_all()
+
+        shard = 0
+        raw_tasks = []
+        for i, (kernel, v) in enumerate(self.tasks.items()):
+            raw_tasks.extend([
+                f'{kernel} {module} {driver_type}\n'
+                for module, driver_type in v
+            ])
+
+            if i < self.get_tps() * (shard + 1):
+                continue
+
+            # filled up the shard, dump it and move on to the next one
+            self._dump_shard(shard, raw_tasks)
+
+            shard += 1
+            raw_tasks = []
+
+        # dump the remaining kernels
+        if len(raw_tasks):
+            self._dump_shard(shard, raw_tasks)
+
+    def match(self, task):
+        if self.regex.match(task.kernel):
+            return True
+        return False
+
+    def append(self, task):
+        if task.kernel not in self.tasks:
+            self.tasks[task.kernel] = [(task.module, task.driver_type)]
+        else:
+            self.tasks[task.kernel].append((task.module, task.driver_type))
+
+    def get_tps(self):
+        if self.shards == 0:
+            return -1
+
+        return len(self.tasks) / self.shards
+
+
+def distribute_shards(total_shards, *builders):
+    total_tasks = 0
+    used_shards = 0
+
+    for builder in builders:
+        total_tasks += len(builder)
+
+    if total_tasks == 0:
+        # No drivers left to build
+        return
+
+    for builder in builders:
+        builder.shards = int(floor((len(builder) * total_shards) / total_tasks))
+        used_shards += builder.shards
+
+    while used_shards < total_shards:
+        min_builder = None
+
+        for builder in builders:
+            if min_builder is None:
+                min_builder = builder
+                continue
+
+            if min_builder.get_tps() > builder.get_tps():
+                min_builder = builder
+
+        min_builder.shards += 1
+        used_shards += 1
+
+
+def main(task_file):
+    # suse4_kernels = r'(?:4\.\d+\.\d+-\d+\.\d+-default)'
+    fc36_kernels = r'(?:5\.[1-9]\d+\..*)'
+    rhel8_kernels = r'(?:(?:4|5)\.\d+\..*)'
+    rhel7_kernels = r'(?:3\.\d+\..*)'
+
+    fc36 = Builder('fc36', fr'^{fc36_kernels}')
```

Comment:
Maybe we could have a dictionary of builders. That way when we a new builder we don't have to add any other code. This could be done in a future PR.

Reply:
Done

Reply:
You could pass that list to distribute_shards and use the list for dump().

Reply:
Also done 🙂